### PR TITLE
Add OfferManager

### PR DESCRIPTION
**Pull request by thomash-acinq**
Add `OfferManager` that handles offer related onion messages (which is all the onion messages that we support).
It can request an invoice and pay it, answer to invoice requests.

Depends on #621 and #623

### CODE REVIEW

**Comment by dpad85** [comment]
Assuming this PR stays in review for a while, and since it's a large change, we could use a temporary specific version name to differentiate this offer branch from the `master` one (they currently both use the same 1.6.2-SNAPSHOT version). It would make it easier to test on Phoenix's end.

**Comment by t-bast** [comment]
Can you rebase this? It will make it easier to test with the latest Bolt 12 changes available on `master`.

File: src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        pendingInvoiceRequests[replyPathId] = PendingInvoiceRequest(payOffer, payerKey, request, currentTimestampSeconds())
+        val numHopsToAdd = max(0, payOffer.minReplyPathHops - 1)
+        val replyPathHops = (listOf(remoteNodeId) + List(numHopsToAdd) { nodeParams.nodeId }).map { IntermediateNode(it) }
+        val lastHop = Destination.Recipient(nodeParams.nodeId, replyPathId)
+        val replyPath = OnionMessages.buildRoute(randomKey(), replyPathHops, lastHop)
+        val messageContent = TlvStream(OnionMessagePayloadTlv.ReplyPath(replyPath), OnionMessagePayloadTlv.InvoiceRequest(request.records))
+        return payOffer.offer.contactInfos.mapNotNull { contactInfo -> buildMessage(randomKey(), randomKey(), listOf(IntermediateNode(remoteNodeId)), Destination(contactInfo), messageContent).right }
+    }
+
+    fun receiveMessage(msg: OnionMessage, channels: Map<ByteVector32, ChannelState>, currentBlockHeight: Int): OnionMessageAction? {
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt:60]: It's overkill to pass the full channels map, the only thing you need is the list of remote `channel_update`s.

File: src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                    )).write().toByteVector()
+                    val blindedRoute = RouteBlinding.create(randomKey(), listOf(remoteNodeId, nodeParams.nodeId), listOf(remoteNodePayload, recipientPayload)).route
+                    val path = Bolt12Invoice.Companion.PaymentBlindedContactInfo(OfferTypes.ContactInfo.BlindedPath(blindedRoute), paymentInfo)
+                    val invoice = Bolt12Invoice(request, preimage, decrypted.blindedPrivateKey, nodeParams.bolt12invoiceExpiry.inWholeSeconds, nodeParams.features.bolt12Features(), listOf(path))
+                    return when (val invoiceMessage = buildMessage(randomKey(), randomKey(), listOf(IntermediateNode(remoteNodeId)), Destination.BlindedPath(decrypted.content.replyPath), TlvStream(OnionMessagePayloadTlv.Invoice(invoice.records)))) {
+                        is Left -> null
+                        is Right -> OnionMessageAction.SendMessage(invoiceMessage.value)
+                    }
+                } else {
+                    // TODO: send back invoice error
```
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt:121]: We should log something whenever we return `null` in this function to help debug.

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt [hunk 3]
```diff
@@ -1240,6 +1249,26 @@ class Peer(
...
             }
+            is PayOffer -> {
+                val invoiceRequests = offerManager.requestInvoice(cmd)
+                invoiceRequests.forEach { input.send(SendMessage(it)) }
+                if (invoiceRequests.isEmpty()) {
+                    _eventsFlow.emit(OfferNotPaid(cmd, OfferPaymentFailure.NoResponse))
+                }
+            }
+            is SendMessage -> peerConnection?.send(cmd.message)
+            // TODO: timeout invoice requests
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt:1290]: Done

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt [hunk 4]
```diff
@@ -74,9 +76,12 @@ data class PayToOpenResponseCommand(val payToOpenResponse: PayToOpenResponse) :
 data class SendPayment(val paymentId: UUID, val amount: MilliSatoshi, val recipient: PublicKey, val paymentRequest: PaymentRequest, val trampolineFeesOverride: List<TrampolineFees>? = null) : PaymentCommand() {
     val paymentHash: ByteVector32 = paymentRequest.paymentHash
 }
+data class PayOffer(val paymentId: UUID, val amount: MilliSatoshi, val quantity: Long, val offer: OfferTypes.Offer, val minReplyPathHops: Int, val trampolineFeesOverride: List<TrampolineFees>? = null) : PaymentCommand()
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt]: Done

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt [hunk 5]
```diff
@@ -86,6 +91,8 @@ sealed class SendPaymentResult : PeerEvent() {
 }
 data class PaymentNotSent(override val request: SendPayment, val reason: OutgoingPaymentFailure) : SendPaymentResult()
 data class PaymentSent(override val request: SendPayment, val payment: LightningOutgoingPayment) : SendPaymentResult()
+data class OfferNotPaid(val request: PayOffer, val reason: OfferPaymentFailure) : PeerEvent()
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt]: Done

File: src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    val remoteNodeId: PublicKey = walletParams.trampolineNode.id
+    private val pendingInvoiceRequests: HashMap<ByteVector32, PendingInvoiceRequest> = HashMap()
+    private val localOffers: HashMap<ByteVector32, OfferTypes.Offer> = HashMap()
+
+    fun registerOffer(offer: OfferTypes.Offer, pathId: ByteVector32) {
+        localOffers[pathId] = offer
+    }
+
+    fun requestInvoice(payOffer: PayOffer): List<OnionMessage> {
+        val payerKey = randomKey()
```
[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/OfferManager.kt]: I've added it to `LightningOutgoingPayment.Details.Blinded` and it's now provided in `PayOffer`.

File: src/commonMain/kotlin/fr/acinq/lightning/NodeParams.kt [hunk 7]
```diff
@@ -151,6 +152,7 @@ data class NodeParams(
     val zeroConfPeers: Set<PublicKey>,
     val liquidityPolicy: MutableStateFlow<LiquidityPolicy>,
     val minFinalCltvExpiryDelta: CltvExpiryDelta,
+    val maxFinalCltvExpiryDelta: CltvExpiryDelta,
```
[the remark this answers is not in the document]
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/NodeParams.kt:155]: Good point! Let's keep it then, with a sane default value exposed by `lightning-kmp` (I don't expect wallet apps to change it from the default).

[the remark this answers is not in the document]
Comment (thomash-acinq) [reply src/commonMain/kotlin/fr/acinq/lightning/NodeParams.kt:155]: I've set it to 360. By default in eclair, we randomly add from 150 to 350 when sending a payment.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt [hunk 8]
```diff
@@ -283,24 +280,25 @@ data class LightningOutgoingPayment(
...
         abstract val paymentHash: ByteVector32
 
         /** A normal lightning payment. */
-        data class Normal(val paymentRequest: PaymentRequest) : Details() {
+        data class Normal(override val paymentRequest: Bolt11Invoice) : Details() {
             override val paymentHash: ByteVector32 = paymentRequest.paymentHash
         }
 
-        /** KeySend payments are spontaneous donations that don't need an invoice from the recipient. */
-        data class KeySend(val preimage: ByteVector32) : Details() {
```
Comment (thomash-acinq) [inline src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:294]: KeySend doesn't seem to be supported, should we just remove all references to it?

**Review by t-bast — COMMENTED** [review]
Looks mostly good to me, I iterated in #633 to add some improvements.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt (hunk 8 above)
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:294]: True, we defined the types initially to ensure that the architecture allowed different types of lightning payments, but we don't support it. You can remove all references to it now that we have the `Blinded` case that shows a different type of lightning payment!

**Comment by t-bast** [comment]
After converging on #633, we should rebase to include #626 and register that default offer:

```kotlin
private val offerManager = run {
  val (pathId, offer) nodeParams.defaultOffer(walletParams.trampolineNode)
  val manager = OfferManager(nodeParams, walletParams, _eventsFlow, logger)
  manager.registerOffer(offer, pathId)
  manager
}
```

**Comment by thomash-acinq** [comment]
I've merged your refactoring, implicit node id and default offer.

**Review by t-bast — APPROVED** [review]

**Pull request merged**